### PR TITLE
Add `config` and `defaultConfig` to `Template`, `Category`, and `Setting`.

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
This PR was made following [this discussion on Discord](https://discord.com/channels/210394599246659585/381219559094616064/1254373736971243542).

The basic idea is that the MCM syntax should be optimized for the most common use-case: a mod has a `config` and `defaultConfig`, and all `Setting` components of the MCM are going to have a `TableVariable` that governs one key in the config.
In code, an example setup is as follows:
```lua
local defaultConfig = {
    a = 1, b = 2, c = 3, d = 4,
    subConfig = { e = 5, f = 6},
}

local config = mwse.loadConfig("test mod", defaultConfig)
```
More often than not, the only thing that changes between `Setting`s would be the `configKey` they govern. So, it  would be nice if we could write
```lua
local template = mwse.mcm.createTemplate{label = "Test Mod", 
    config = config, 
    defaultConfig = defaultConfig
}
local page = template:createSideBarPage{label = "Page"}

page:createSlider{label = "a", configKey = "a"}
page:createSlider{label = "b", configKey = "b"}
page:createSlider{label = "c", configKey = "c"}
page:createSlider{label = "d", configKey = "d"}
```
In this "ideal world", the above code would automatically create a `TableVariable` for each slider, and initialize the `defaultSetting` of that slider to the value stored in `defaultConfig`. In other words, the above code would be equivalent to:
```lua
local template = mwse.mcm.createTemplate{label = "Test Mod"}

local page = template:createSideBarPage{label = "Page"}

page:createSlider{label = "a", 
    variable = mwse.mcm.createTableVariable{
        id = "a",
        table = config,
        defaultSetting = defaultConfig.a
    }
}
page:createSlider{label = "b", 
    variable = mwse.mcm.createTableVariable{
        id = "b",
        table = config,
        defaultSetting = defaultConfig.b
    }
}
page:createSlider{label = "c", 
    variable = mwse.mcm.createTableVariable{
        id = "c",
        table = config,
        defaultSetting = defaultConfig.c
    }
}
page:createSlider{label = "d", 
    variable = mwse.mcm.createTableVariable{
        id = "d",
        table = config,
        defaultSetting = defaultConfig.d
    }
}
```

This PR brings us closer to that "ideal world". It works as follows:
- When creating a `Template`, you can pass a `config` and `defaultConfig`.
- When creating a `Category` (and thus any `Page`), the `config` and/or `defaultConfig` will be inherited from the `parentComponent`
  - These can be overridden by explicitly specifying another `config` and `defaultConfig`. This is to make this new syntax compatible with configs that have nested subtables.
- When creating a `Setting`:
  - If a `variable` was specified explicitly:
    - Do not repleace it. But instead, see if we can use `parentComponent.defaultConfig` and `variable.id` to add a `defaultSetting` to that variable, if no other `defaultSetting` is present.
  - If no `variable` was specified explicitly:
    - If possible, use `parentComponent.config`, and `self.configKey` to create a new `TableVariable`.
      -  This is done by setting `id = self.configKey`,  `table = parentComponent.config`, and if possible, `defaultSetting = parentComponent.defaultConfig[self.configKey]`
      - Here, `self` refers to the new `Setting` being constructed, and `parentComponent` refers to the `Category`/`Page` that the setting lives in.

The above description is a rough overview of the supported functionality. The actual implementation in this PR is much more flexible. But I'm not sure if the extra flexibility is actually useful, or if it just makes things more confusing. That's something I would like feedback on. In particular:
- You can specify a `configKey` when creating a `Category`. This is implemented so that the following two constructors are identical: 
```lua
-- Both of these do the same thing.
page:createCategory{ configKey = "subConfig" }
page:createCategory{ config = page.config.subConfig, defaultConfig = page.defaultConfig.subConfig}
```
- The motivation for this is that it should help to minimize errors that result from `config` and `defaultConfig` becoming out-of-sync with each other. But I think it would be clunky to support both of the above approaches. I'm just not sure which one to pick.
  - You can also directly specify a `config` and `defaultConfig` when creating a `Setting`. This would override the `config` and `defaultConfig` values stored in the `parentComponent`, but only for a specific `Setting`. I'm not too happy with this though, since I think it just makes things more confusing. In this case, it would probably be better to just create a `TableVariable` manually.

I'm open to any feedback/suggestions on how the concept could be improved. It's in no way finalized.

I generated some of the documentation, to help explain the concept. But the included documentation should be treated as a (very) rough draft. And I still need to document that `create<Component>` functions as well.

### CODE REVIEW

File: misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua [hunk 1]
```diff
@@ -21,10 +23,38 @@ Setting.restartRequiredMessage = mwse.mcm.i18n("The game must be restarted befor
...
+		local defaultConfig = data.defaultConfig or parent and parent.defaultConfig
+
+		-- Get the default setting. Include `nil` checks so we can handle it being `false`.
+		local defaultSetting = data.variable and data.variable.defaultSetting
+		if defaultSetting == nil then
+			defaultSetting = data.defaultSetting
+			if defaultSetting == nil and defaultConfig and configKey then
+				defaultSetting = defaultConfig[configKey]
+			end
+		end
```
Comment (C3pa) [inline misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:40]: IMHO, could be a bit more readable as:
```lua
local defaultSetting = data.variable and data.variable.defaultSetting or data.defaultSetting
if defaultSetting == nil and defaultConfig and configKey then
		defaultSetting = defaultConfig[configKey]
end
```

**Review by C3pa — APPROVED** [review]
Looks good. I'll just have to test this out before merge.

File: misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua (hunk 1 above)
Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:40]: I would have preferred that as well, but the downside of that approach is that it has some weird behavior when `data.variable.defaultSetting == false` (e.g. if `data` is a `YesNoButton`). More specifically, if `data.defaultSetting == nil` and `data.variable.defaultSetting == false`, then 
```lua
local defaultSetting = data.variable and data.variable.defaultSetting or data.defaultSetting
```
will set `defaultSetting = nil`.

**Comment by herbertgithubaccount** [comment]
> Looks good. I'll just have to test this out before merge.

Should I get started on writing the rest of the documentation? Or do you think I should make some changes to the API first?

**Comment by C3pa** [comment]
The API looks good to me. Feel free to add the rest of the docs.

As I said, I'll test this out locally and see if I can find some weird edge cases.

File: misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua (hunk 1 above)
Comment (C3pa) [reply misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:40]: Oh, right, it's not functionally identical. Still, my inner non-nester feels it's a bit nicer if it's done as:
```lua
local defaultSetting = data.variable and data.variable.defaultSetting
if defaultSetting == nil then
	defaultSetting = data.defaultSetting
end

if defaultSetting == nil and defaultConfig and configKey then
	defaultSetting = defaultConfig[configKey]
end
```
Feel free to ignore this suggestion of you don't find it more readable that way.

Comment (herbertgithubaccount) [reply misc/package/Data Files/MWSE/core/mcm/components/settings/Setting.lua:40]: I'll make this change. I agree that it's more readable this way. It's always nice to cut down on nesting whenever possible.

**Comment by herbertgithubaccount** [comment]
Sounds good. I'll get started on the rest of the documentation. And I make documentation changes later if other things come up.

> Unrelated, if I got the plan right, we'll first get this in, and then we'll see what to do in https://github.com/MWSE/MWSE/pull/542?

I'm also a bit confused about the plan, but I think that what you outlined would be a good approach. #542 would be a lot easier to implement after this gets merged, since the `defaultConfig` will already be included in the relevant `Component`s. It would also mean we don't have to worry about using a heuristic, which would make it more reliable as well.

**Comment by C3pa** [comment]
All right, I used this to test:
```lua
local defaultConfig = {
	a = 1, b = 2, c = 3, d = 4,
	subConfig = { e = 5, f = 6},
}

local config = mwse.loadConfig("A test mod", defaultConfig)

event.register("modConfigReady", function()	
	local template = mwse.mcm.createTemplate{label = "Experimental MCM", 
	config = config, 
	defaultConfig = defaultConfig
	}
	template:register()
	local page = template:createSideBarPage{label = "Page"}

	page:createSlider{label = "a", configKey = "a"}
	page:createSlider{label = "b", configKey = "b"}
	page:createSlider{label = "c", configKey = "c"}
	page:createSlider{label = "d", configKey = "d"}

	local category = page:createCategory{label = "subConfig", configKey = "subConfig" }
	category:createSlider{label = "e", configKey = "e"}
	category:createSlider{label = "f", configKey = "f"}
end)
```
I didn't find anything obviously broken, and other mods MCM's worked okay.

**Comment by C3pa** [comment]
What I found is that by running https://github.com/MWSE/MWSE/blob/master/docs/.vscode/tasks.json, which can be run when `docs\` folder is opened in VSCode, I got the following output:
```
WARNING -  Doc file 'types/mwseMCMCategory.md'
           contains a link
           './namedTypes/mwseMCMTableVariable.md', 
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files. 
WARNING -  Doc file 'types/mwseMCMCategory.md'
           contains a link
           './namedTypes/mwseMCMSetting.md', but   
           the target
           'types/namedTypes/mwseMCMSetting.md' is 
           not found among documentation files.    
WARNING -  Doc file 'types/mwseMCMCategory.md'     
           contains a link
           './namedTypes/mwseMCMTableVariable.md', 
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files. 
WARNING -  Doc file 'types/mwseMCMSetting.md'
           contains a link
           './namedTypes/mwseMCMTableVariable.md', 
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files. 
WARNING -  Doc file 'types/mwseMCMSetting.md'
           contains a link
           './namedTypes/mwseMCMTableVariable.md', 
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files. 
WARNING -  Doc file 'types/mwseMCMTemplate.md'
           contains a link
           './namedTypes/mwseMCMTableVariable.md', 
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files.
WARNING -  Doc file 'types/mwseMCMTemplate.md'
           contains a link
           './namedTypes/mwseMCMSetting.md', but
           the target
           'types/namedTypes/mwseMCMSetting.md' is
           not found among documentation files.
WARNING -  Doc file 'types/mwseMCMTemplate.md'
           contains a link
           './namedTypes/mwseMCMTableVariable.md',
           but the target
           'types/namedTypes/mwseMCMTableVariable.md'
           is not found among documentation files.
```
So, looks like some relative links are incorrect. Can you just fix these before merge?

**Comment by herbertgithubaccount** [comment]
I found out i got the type links wrong in a previous version but I thought it was fixed in the most recent commit. Maybe my local version of out of sync? I couldn't find any mention of `namedTypes/mwseMCMSetting.md` when searching through the files on my local version, and I don't seem to get those warning when I try to build my local version either. But I could definitely be missing something.

**Comment by C3pa** [comment]
Oh, my bad, I didn't pull the latest 2 commits. There are no warnings with them.

**Comment by C3pa** [comment]
Thanks for working on this Herbert!

**Comment by herbertgithubaccount** [comment]
> Oh, my bad, I didn't pull the latest 2 commits. There are no warnings with them.

No worries! I'm just glad my local files didn't get out of sync. Thanks for the help on this PR and all the others.

File: autocomplete/definitions/namedTypes/mwseMCMTemplate/new.lua [hunk 2]
```diff
@@ -7,6 +7,14 @@ return {
 		tableParams = {
 			{ name = "name", type = "string", optional = true, description = "The name field is the mod name, used to register the MCM, and is displayed in the mod list on the lefthand pane." },
 			{ name = "label", type = "string", optional = true, description = "Used in place of `name` if that argument isn't passed. You need to pass at least one of the `name` and `label` arguments. If `headerImagePath` is not passed, a UI element will be created with `label` as text." },
+			{ name = "config", type = "table", optional = true, 
+				description = "Stores a config that should be used by this mods `Setting`s. Sub-configs can be accessed by passing a `configKey` to any `Page`s nested inside this template. \z
+					If provided, this config will be used to generate [`mwseMCMTableVariable`s](./mwseMCMTableVariable.md) for the  any [`mwseMCMSetting`s](./mwseMCMSetting.md) made inside this template."
+			},
+			{ name = "defaultConfig", type = "table", optional = true,
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/mwseMCMTemplate/new.lua:14]: Looks like the `config` and `defaultConfig` fields weren't added to the mwseMCMTemplate definition.
